clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -734,11 +745,11 @@ private ClusterResourceListeners configureClusterResourceListeners(Serializer<K>
     private int partition(ProducerRecord<K, V> record, byte[] serializedKey, byte[] serializedValue, Cluster cluster) {
         Integer partition = record.partition();
         if (partition != null) {
-            List<PartitionInfo> partitions = cluster.partitionsForTopic(record.topic());
-            int lastPartition = partitions.size() - 1;
+            int lastPartition = cluster.partitionsForTopic(record.topic()).size() - 1;
             // they have given us a partition, use it
-            if (partition < 0 || partition > lastPartition) {
-                throw new IllegalArgumentException(String.format("Invalid partition given with record: %d is not in the range [0...%d].", partition, lastPartition));
+            if (partition > lastPartition) {
```

Comment:
I wonder if it would be better to fail in `waitOnMetadata` instead of having the logic in two places.